go/workflow.go

```diff
@@ -0,0 +1,333 @@
+package loyalty
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"go.temporal.io/sdk/temporal"
+	"go.temporal.io/sdk/workflow"
+)
+
+const TaskQueue = "CustomerLoyaltyTaskQueue"
+const EventsThreshold = 10000
+
+// Signal, query, and error string constants
+const (
+	SignalCancelAccount       = "cancelAccount"
+	SignalAddPoints           = "addLoyaltyPoints"
+	SignalInviteGuest         = "inviteGuest"
+	SignalEnsureMinimumStatus = "ensureMinimumStatus"
+	QueryGetStatus            = "getStatus"
+	QueryGetGuests            = "getGuests"
+)
+
+const (
+	EmailWelcome            = "Welcome to our loyalty program! You're starting out at '%v' status."
+	EmailGuestCanceled      = "Sorry, your guest has already canceled their account."
+	EmailGuestInvited       = "Congratulations! Your guest has been invited!"
+	EmailInsufficientPoints = "Sorry, you need to earn more points to invite more guests!"
+	EmailPromoted           = "Congratulations! You've been promoted to '%v' status!"
+	EmailDemoted            = "Unfortunately, you've lost enough points to bump you down to '%v' status. 😞"
+	EmailCancelAccount      = "Sorry to see you go!"
+)
+
+type GuestAlreadyCanceledError struct {
+	msg string
+}
+
+func (e *GuestAlreadyCanceledError) Error() string { return e.msg }
+
+func CustomerLoyaltyWorkflow(ctx workflow.Context, customer CustomerInfo) (err error) {
+	logger := workflow.GetLogger(ctx)
+	logger.Info("Loyalty workflow started.", "CustomerInfo", customer)
+
+	ao := workflow.ActivityOptions{
+		StartToCloseTimeout: 10 * time.Second,
+	}
+	lao := workflow.LocalActivityOptions{
+		ScheduleToCloseTimeout: 10 * time.Second,
+		RetryPolicy: &temporal.RetryPolicy{
+			NonRetryableErrorTypes: []string{"GuestAlreadyCanceledError"},
+		},
+	}
+	ctx = workflow.WithActivityOptions(ctx, ao)
+	ctx = workflow.WithLocalActivityOptions(ctx, lao)
+
+	info := workflow.GetInfo(ctx)
+	selector := workflow.NewSelector(ctx)
+	activities := Activities{}
+	workflowCanceled := false
+	var errSignal error
+
+	logger.Debug("Got workflow info.", "WorkflowID", info.WorkflowExecution.ID)
+
+	logger.Info("Validating customer info.")
+	validateCustomerInfo(&customer)
+
+	if info.ContinuedExecutionRunID == "" {
+		err = workflow.ExecuteActivity(ctx, activities.SendEmail,
+			fmt.Sprintf(EmailWelcome, customer.StatusLevel.Name)).
+			Get(ctx, nil)
```

Comment:
Should note on all of these emails, you are going to just hang the workflow forever if the email is failing to send. I don't see a problem with it, especially if email is a required part of this, just letting you know. Some people only try to send an email so many times (and increase the backoff between each from the default of rapid repeat) and don't let it stop down the workflow.

Reply:
👍 removed failing the workflow on SendEmail activity failure, and set a max attempts.